### ISSUE TITLE
Soft Morphological operations : 25 Structuring Elements -> 1 radius

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#76]
<html>
<body>
<html><head></head><body><h1>Note: Soft Morphology — Design Decisions &amp; Understanding</h1>
<h2>Context</h2>
<p>Refactoring <code>DiffDilation</code>, <code>DiffErosion</code>, <code>DiffOpening</code>, <code>DiffClosing</code> from 25 independent learnable SE weights (<code>se_values</code>) to a single learnable scalar <code>radius</code>.</p>
<hr>
<h2>Why soft morphology is still morphology</h2>
<p>Classical morphology uses binary SE (0/1). Soft morphology replaces it with continuous values (0.0–1.0) — a <strong>differentiable relaxation</strong>. As values converge toward binary and temperature increases, it recovers exact classical morphology. This is academically well-established. The learned circular SE from a uniform initialization is evidence it works correctly.</p>
<hr>
<h2>Why <code>radius</code> instead of 25 independent weights</h2>
<ul>
<li>25 params is an outlier in the system (all other ops have 1–3 params)</li>
<li>Independent weights can learn nonsensical (non-circular) shapes</li>
<li>A single <code>radius</code> encodes the entire disk shape geometrically — more meaningful and minimal</li>
<li>Paper story is cleaner: one scalar controls a spatially meaningful shape</li>
</ul>
<hr>
<h2>How the SE is generated from <code>radius</code></h2>
<p><strong>Step 1 — precompute distance grid</strong> (fixed, never changes):</p>
<pre><code>dist[i,j] = sqrt(x² + y²)   ← euclidean distance from kernel center
</code></pre>
<p>For 5×5 kernel, center is (2,2). Each cell gets its distance from center.</p>
<p><strong>Step 2 — generate soft disk via sigmoid:</strong></p>
<pre><code>se[i,j] = sigmoid( alpha × (radius - dist[i,j]) )
</code></pre>
<ul>
<li><code>radius - dist &gt; 0</code> → inside disk → sigmoid → close to 1 (active)</li>
<li><code>radius - dist &lt; 0</code> → outside disk → sigmoid → close to 0 (inactive)</li>
<li><code>alpha = 5.0</code> (fixed) controls sharpness of the disk boundary</li>
</ul>
<p><strong>Step 3 — optimizer learns the right disk size:</strong></p>
<ul>
<li>Start: <code>radius = 2.0</code> (middle of valid range, gives optimizer room both ways)</li>
<li>Each training step: gradient flows back through sigmoid → updates radius</li>
<li>Result: radius converges to the disk size that minimizes loss for the task</li>
</ul>
<hr>
<h2>Clamping radius</h2>
<p><code>radius</code> must stay in <code>[0.1, 3.5]</code>:</p>
<ul>
<li><code>&lt; 0.1</code> → SE nearly empty, morphology has no effect</li>
<li><code>&gt; 3.5</code> → SE fully saturated, no shape discrimination (max dist in 5×5 ≈ 2.83)</li>
</ul>
<h3>Two types of clamp — not interchangeable</h3>

  | torch.clamp(self.radius, 0.1, 3.5) | self.radius.data.clamp_(0.1, 3.5)
-- | -- | --
Inside autograd graph | ✅ Yes | ❌ No
Use in forward() | ✅ Correct | ❌ Wrong
Use after optimizer.step() | unnecessary | ✅ Safe correction

<p>Always use <code>torch.clamp</code> inside <code>forward()</code> / <code>_get_se()</code> so gradient flows correctly.</p>
<hr>
<h2>Key implementation details</h2>
<pre><code class="language-python"># __init__
K = 5
coords = torch.arange(K, dtype=torch.float32) - (K // 2)
grid_y, grid_x = torch.meshgrid(coords, coords, indexing='ij')
self.register_buffer('dist', torch.sqrt(grid_x**2 + grid_y**2))
self.radius = nn.Parameter(torch.tensor(2.0))
self.alpha = 5.0  # fixed, not learned

# _get_se() inside forward()
radius = torch.clamp(self.radius, 0.1, 3.5)
se = torch.sigmoid(self.alpha * (radius - self.dist))
</code></pre>
<p><code>dist</code> is a <strong>buffer</strong> — moves to GPU with the model, no gradient.</p>
<hr>
<h2>Open design question</h2>
<p><code>DiffOpening</code> (Erosion → Dilation) and <code>DiffClosing</code> (Dilation → Erosion) — does each internal op get its <strong>own</strong> <code>radius</code>, or do they <strong>share</strong> one? Needs explicit decision before finalizing implementation.</p>
<hr>
<h2>LLM prompt update</h2>
<p>Replace:</p>
<pre><code>se_values: 25 floats for 5x5 structuring element, all start at 1.0
</code></pre>
<p>With:</p>
<pre><code>radius: single float, range [0.5, 3.0], default 2.0
</code></pre>
<p>Expected LLM output after refactor:</p>
<pre><code class="language-python">('DiffDilation', [-1], {'radius': [2.0]})
</code></pre></body></html>
</body>
</html>

**Comment by xp1632** [comment]
https://claude.ai/chat/852c6507-84f0-4907-abc0-eaa6b324624b

**Comment by xp1632** [comment]
### how to state in paper

## In the Method section

**Structuring Element parameterization:**

> Each morphological operation is parameterized by a single learnable scalar $r$ representing the disk radius. The structuring element $\mathbf{S} \in \mathbb{R}^{K \times K}$ is generated analytically as:
>
> $$S_{ij} = \sigma\!\left(\alpha \cdot (r - d_{ij})\right)$$
>
> where $d_{ij} = \sqrt{(i-c)^2 + (j-c)^2}$ is the Euclidean distance from kernel center $c$, and $\alpha = 5.0$ is a fixed sharpness constant. Dilation is approximated as:
>
> $$(\mathbf{I} \oplus \mathbf{S})_{xy} = \max_{(i,j):\ S_{ij} > \epsilon}\ I_{x+i,\ y+j}$$
>
> with inactive positions ($S_{ij} \approx 0$) excluded from the max via masking. Erosion follows symmetrically with min.

---

## In the Discussion / Design Choices

One sentence is enough:

> This formulation reduces each morphological operation to a single optimizable parameter while preserving geometric interpretability — the optimizer learns the appropriate disk size rather than 25 independent unconstrained weights.

---

## What NOT to claim

- Don't claim this is novel — soft morphology is established, cite prior work (e.g. Masci et al. 2012, Shen & Sandham 2004)
- Frame it as a **design choice for parameter efficiency**, not a theoretical contribution
- The contribution is using it *within an LLM-guided optimization pipeline*, not the soft morphology itself
